Add control structures and pattern matching

lbseale opened the pull request:
Control Structures lesson covers

- if statements
- case statements
- guards

Pattern Matching 

- Covers matching plain data constructors, lists, Maybe, Either
- Assumes that all these types have been introduced already

lbseale committed:
Add control structures and pattern matching

TrueBoxGuy commented on en/lessons/basics/control-structures.md:
```diff
@@ -1,207 +1,237 @@
+  where
+    tmpInC = (tmpInF - 32.0) * ( 5.0 / 9.0 )
+
+-- Clothing Sizer
+-- Note that `yourSize` is defined in a where clause, then used in the
+-- definition
+whatSize :: Int -> Text
+whatSize heightInCm = "You're a size " <> yourSize
+  where
+    yourSize
+        | heightInCm > 0   && heightInCm <= 166 = "Small"
+        | heightInCm > 166 && heightInCm <= 178 = "Medium"
+        | heightInCm > 178                      = "Large"
+        | otherwise                             = "Not sure"
```
I like that this example shows another usage pattern for guards. 

This may be a moot point, but I don't feel that the usage pattern here matches this example, and this example is somewhat similar to the last one. I think that a more categorical example with numerical output (e.g. one which calculates scores in a game like the Prisoner's Dilemma and displays them) would be more suitable and interesting.

lbseale replied:
Agreed, it's not a big difference from the other example. I like your game-scoring idea, I'll think about an example for that

TrueBoxGuy commented on en/lessons/basics/pattern-matching.md:
```diff
@@ -1,107 +1,268 @@
-{2, 1}
-iex> x
-2
+#### Examples of Not Patterns
```
```suggestion
#### Examples of Non-Patterns
```

TrueBoxGuy commented on en/lessons/basics/pattern-matching.md:
```diff
@@ -1,107 +1,268 @@
+7        -- An `Int` value
+[]       -- An empty list
+(x:xs)   -- A list with one or more values, constructed with `:`
+         -- `x` is the first element of the list, `xs` is the rest of the list
```
```suggestion
         -- `x` is the first element of the list, and `xs` is the rest of the list
```

TrueBoxGuy commented on en/lessons/basics/pattern-matching.md:
```diff
@@ -1,107 +1,268 @@
+A pattern is a combination of data constructors that could match a value
 
-## Pin Operator
+Functions are not permitted in patterns -- If you need to use a function,
```
I think such strong emphasis on functions not being allowed in patterns may be mildly harmful, given that view patterns and pattern synonyms exist. 

It's probably inappropriate to talk about  these language extensions here, but it may be beneficial to rephrase your explanation in terms of constructors.

lbseale replied:
I did not know about this language extension, so I'm learning as well. I honestly wanted a better diversity of counter-examples but I couldn't think of more than these little function applications. 

Can you think of anything other than a function application to use as a counter-example?

TrueBoxGuy replied:
I really can't think of anything right now, but maybe something will pop into my head in the future when I'm coding. 

I think `++` is a useful counter-example (even though it's still function application).

TrueBoxGuy commented on en/lessons/basics/pattern-matching.md:
```diff
@@ -1,107 +1,268 @@
+(7 - 2)     -- Functions are not allowed in patterns, `-` is a function
+(x > 5)     -- Functions are not allowed in patterns, `>` is a function
+            -- Consider a guard
+(=="apple") -- Seriously, no functions allowed, `==` is a function
```
```suggestion
(7 - 2)     -- Functions are not allowed in patterns: `-` is a function
(x > 5)     -- Functions are not allowed in patterns: `>` is a function
            -- Consider a guard
(=="apple") -- Seriously, no functions allowed: `==` is a function
```

TrueBoxGuy commented on en/lessons/basics/pattern-matching.md:
```diff
@@ -1,107 +1,268 @@
-"world"
-iex> %{^key => value} = %{:hello => "world"}
-** (MatchError) no match of right hand side value: %{hello: "world"}
+Using pattern matching to handle lists is very common, let's see some examples
```
```suggestion
Using pattern matching to handle lists is very common. Let's see some examples
```

TrueBoxGuy reviewed (commented):
This looks great, and covers a lot of Haskell School content!

I myself don't have much to say. One pattern I may consider adding is the usage of both pattern matching and guards. Here's a contrived example of what I mean:
```hs
checkPositive (Just x) | x >= 0 = Just x
checkPositive _ = Nothing
```

lbseale committed:
Apply suggestions from code review

Co-authored-by: TrueBoxGuy <17554188+TrueBoxGuy@users.noreply.github.com>

lbseale committed:
Add some new examples for better coverage

- Replace clothing sizer example with prisoner's dilemma example of
guards used in a `where` clause
- Add an example showing both pattern matching and guards in the same
function
- Thanks to TrueBoxGuy for these suggestions

commit d8ba4e0139f7cd12370d3cab25c7ca25e01977be
Author: Luke <19484084+lbseale@users.noreply.github.com>
Date:   Tue Sep 28 06:20:34 2021 -0700

    Remove ControlExamples.hs test file

commit a88af92bc4f954df2aeec1c46020be1ab9a0bd52
Author: Luke <19484084+lbseale@users.noreply.github.com>
Date:   Tue Sep 28 06:19:52 2021 -0700

    Add an example of guards combined with patterns

commit 81295e35064d4965c2fe31c5d441cd59d6745ae0
Author: Luke <19484084+lbseale@users.noreply.github.com>
Date:   Mon Sep 27 07:08:14 2021 -0700

    Add prisoners dilemma example

    - Get rid of clothing sizer example, it's too similar to the other one
    - Still need to add a pattern-matching + guards example

Kleidukos commented on en/lessons/basics/control-structures.md:
```diff
@@ -1,207 +1,246 @@
+- The expression between the `case` and `of` keywords is evaluated
+- Its result is looked up in the list of patterns
+- For the first pattern that matches, its expression is evaluated
+- The underscore (`_`) is a catch-all pattern that will always match
```
I would love a tiny addition, like a Note, to say that if you put the `_` somewhere, every pattern that comes *after* will be ignored. Highlight that order of patterns matters!

Kleidukos reviewed (changes requested):
Thank you very much @lbseale! I'll pass the review to @TrueBoxGuy for the last part of the review and I think we can merge it afterwards ;-)

why-not-try-calmer commented on en/lessons/basics/control-structures.md:
````diff
@@ -1,207 +1,246 @@
+```haskell
+{-# LANGUAGE OverloadedStrings #-}
+
+import Data.Text ( Text )
````
Wouldn't it be preferable to keep inside the Prelude namespace as much as possible? I know `String` are bad for many things, but don't they offer a simple and intuitive interface for teaching to new users?

Kleidukos replied:
Reasoning about lazy linked lists when speaking of text is far from being intuitive.

why-not-try-calmer replied:
So this document is solely about syntax, and the syntax/interface provided by Strings is extremely intuitive. It's only their semantics that is not. But since the semantics of Haskell in general is not intuitive at all, why not handle semantic difficulties separately?

Kleidukos replied:
Because then you have to undo the intuition that you promote afterwards. We already have enough space leak issues with people blindingly using `String` because they are "so intuitive to use", I don't want to contribute legitimising `String` as a valid textual datatype: It is not and we should stop fooling ourselves.

lbseale replied:
I noticed that the very first lesson uses `Text`, and I figured that was intentional. So I did the same here. Also using the `<>` operator is just as intuitive to me as using `++` with `String`

why-not-try-calmer replied:
> I noticed that the very first lesson uses `Text`, and I figured that was intentional. So I did the same here. Also using the `<>` operator is just as intuitive to me as using `++` with `String`

I very much understand the idea of promoting the `text` package. Everyone doing anything serious in Haskell uses it instead of `String` for almost anything regarding text contents -- no question about that. Also `Data.Text` might become part of the Prelude one day (after we move from Utf16 to Utf8 probably). But if you look at any lauded pedagogical material for Haskell (_Haskell Programming From First Principles_, more controversially _Get Programming With Haskell_ for example), the `Text` type is introduced _much later_ and for reasons. (quoting from Kurt Manning:

> While learning Haskell, String is useful for two reasons. First, as mentioned, many of the basic string utilities are baked into the standard Prelude. Second, lists are to Haskell what arrays are to C. Many concepts in Haskell are nicely demonstrated with lists, and strings are useful lists. For learning purposes, feel free to stick with String.

To me the first reason means: now you need to `import Data.Text` everywhere, which clutters the material, while the second reason means that you can easily use String to illustrate lists -- including the list monad. I think these are two valid and strong reasons for the sake of pedagogy.

And there are other reasons; now you need to explain why one should import the strict and not the lazy version, or vice-versa; why `Data.Text` has its own separate functions different from the Prelude functions; and if the point is efficient text processing, why `Data.Text` instead of `ut8-bytestring` or `Data.Vector.Unboxed.Vector Char`, and why --- exception! -- `Text.IO` is awfully inefficient and should be avoided.

In other word, it seems to me that using `Text` everywhere is a bit of a premature optimization that opens a can of worms. How about instead:

1) using String whenever the focus is on cons-able data structures?; and
2) using Text whenever the focus is on efficient text processing?

Kleidukos replied:
Well, you're entitled to your opinion. :)

why-not-try-calmer replied:
> Well, you're entitled to your opinion. :)

Can we get the perspective of the folks supporting the Haskell School project on this? It seems to be important, both as a technical detail and as a case illustrating the vision of this project. It doesn't feel right to discard the dozens of examples of user-friendly books and documentation about Haskell by implicitly telling them that Text is better than String for the purpose of learning Haskell.

TrueBoxGuy replied:
Haskell School does need to keep its own style (as a reference material to get people quickly acquainted with Haskell). 

I don't feel like it's particularly productive to debate over the usage of `Text`: it's a project wide decision and debating over it slows down the progress of individual pull requests. It would be better to get to an initial release and then see whether the benefits of using `Text` have been realised. 

However, these internal decisions perhaps aren't the clearest from the outside and your review has been very helpful in showing potential misunderstandings a reader may have.

why-not-try-calmer replied:
Sure, I don't want to slow down this PR with a project-wide decision. Shall I open an Issue then?

TrueBoxGuy replied:
I think we're considering putting these decisions in the style guide, and then maybe there can be a discussion about the style guide somewhere.

ulysses4ever replied:
@why-not-try-calmer I share the concern of using `text` from the get-go, [as I mentioned in an earlier PR](https://github.com/haskellfoundation/HaskellSchool/pull/7#discussion_r642747500).

I encourage Haskell School government to seek a wider input on this matter, e.g. via Discourse / Reddit / mailing lists.

lbseale replied:
I support teaching `Text` from the beginning. I learned Haskell in the last year, and it feels bad to learn topics as a beginner, only to find out that they're actually bad practice. I think there are several examples of this

1. `String` - you find out that the tools for it are clunky and that you should use `Text` instead
2. Lists - You find out that usually you should use some `Map` type instead, and that lists involve unsafe partial functions you're not supposed to use
3. Explicit recursion - You find out it can cause memory leaks, and that you rarely actually use it in practice

There are probably others. TBH these issues are why I want to contribute to this project.

Kleidukos replied:
I am going to speak as the Editor-in-chief of this content platform: 

Teaching `Text` from the beginning has several advantages:
* It helps reducing cognitive load, as the learner doesn't have to "un-learn" String
* It gives the learner proper habits from the beginning
* It reduces the amount of confusion arising from the fact that `String` is simply an alias for `[Char]` and GHC doesn't always expand or retain aliases.
* The `String` utilities baked into the Prelude are simply insufficient when going beyond very simple or contrived examples, and they possess equivalents in `Text`.

As a general statement, `String` is a historical accident that we kept and I do not wish to promote something from which languages older than Haskell like Erlang have moved away.

@why-not-try-calmer @ulysses4ever You may not agree with this decision but such is life. @TrueBoxGuy is editor for the English language part of Haskell school, and I (and transitively, the executive team of the Haskell Foundation) have put trust in his judgement.

JonathanLorimer replied:
FWIW I am strongly in favour of using `Text` over `String`.

Perhaps it is our fault for not being clear enough in some coding style document, but one of the primary motivating factors of this educational tool is to get a new haskeller up to speed and working with haskell in real world contexts quickly. To this end we did not want to delay learning to use `Text` over `String`. We also don't want to delay introducing them to the critical haskell libraries that they will inevitably need to learn. Also, if the argument is that `String` is convenient for teaching pattern matching, surely any other example of a list that is not an overloaded concept would be an equally good example?

Haskell School seeks to learn from mistakes made by other educational haskell materials (like the ones you mention @why-not-try-calmer). If you don't consider them to be mistakes, then perhaps I can appeal to the virtues of having diversity in available resources.

Finally, I would like to make an appeal to authority. @Kleidukos was designated the shepherd of this project, and TBG / myself have put in hours of time, early on, contributing to the project. All three of us came to the conclusion that using `Text` throughout is the approach we think is best.

I don't think this project should be an autocracy, if @ulysses4ever (who has also been around this project and contributing since the beginning) and `why-not-try-calmer` want to make a case for teaching `String` over `Text` I encourage you to bring it up in the Haskell Foundation slack and we can have it out there. But I hope its clear that this was not a one off decision, it is in service of our larger vision for the project. To summarize I think we should stick with the current approach since:

1. We have made the decision to prioritize getting the learner doing real world things as quickly as possible. This means introducing them to libraries outside of prelude earlier than other pedagogical materials
2. We think that, at best, we are improving on the canon of haskell educational material. At worst, we are providing diversity, which I argue is an end in its own right.
3. This decision was arrived at by people who were charged with executing the project, and who have invested their time in bringing it about. Know that we have discussed this in the past, and our decision was made based on what we understand the mandate of the project to be.

JonathanLorimer replied:
Okay, it seems @Kleidukos and I posted at almost exactly the same time. I hope this doesn't come across as needless dog piling. Just wanted to clarify that.

shadowcat-mst replied:
Representing strings as List-of-Char, especially in a unicode world, is one of the most common initially enticing decisions for all functional programming language implementors that in the long run turn out to be where sanity and performance go to die. 

I think Erlang mostly had it because of starting off being implemented in prolog, which really mostly didn't care about string handling as originally designed.

Any "let's implement a lisp" tutorial that's written by somebody who's worked with non-toy systems will almost certainly contain a paragraph to the effect of "we know this looks like a really tempting idea but trust us, don't".

It's unfortunate that String already stole a bunch of operator names that one might've wished one could use for Text, but that's the nature of long lived systems, and teaching newbies Text first will be a slightly steeper learning curve up front in return for much *much* less pain for them later, so honestly starting them off with String in 2021 would just be unnecessary cruelty.

Yours, mst, speaking from bitter experience over multiple FP implementations.

why-not-try-calmer replied:
Fair enough, this PR is probably not the best place for debating the presupposition -- which by now you'll know I don't accept -- that illustrating basic Haskell with `String`s entails a general endorsement or promotion of `String`s. Feel free to ignore the initial comment that spawned this discussion.

why-not-try-calmer commented on en/lessons/basics/control-structures.md:
````diff
@@ -1,207 +1,246 @@
 ```
 
-The `_` variable is an important inclusion in `case/2` statements. Without it, failure to find a match will raise an error:
+- If the first expression evaluates to `True`, then the expression after the
````
Would it be overdoing things to recall the users with a background in dynamic languages that there is no type coercion here, so that `True` really a value constructed from the 0-place `True` constructor and not some "truthy" value as in Javascript or Python?

Kleidukos replied:
Booleans are already covered in the first chapter: https://github.com/haskellfoundation/HaskellSchool/blob/4df5d05991e720cbc9255685a45388ed8dcf1900/en/lessons/basics/basics.md#booleans

lbseale replied:
I added a small comment to be more explicit about this, and to be consistent with the section on Guards

why-not-try-calmer commented on en/lessons/basics/control-structures.md:
```diff
@@ -1,207 +1,246 @@
-"Will match"
+- The expression between the `case` and `of` keywords is evaluated
+- Its result is looked up in the list of patterns
+- For the first pattern that matches, its expression is evaluated
```
Wouldn't be nice to me more explicit: "The first pattern that matches -- and only it -- has its expression evaluated"?

lbseale replied:
Added to latest push

why-not-try-calmer commented on en/lessons/basics/control-structures.md:
```diff
@@ -1,207 +1,246 @@
-
-The special form `with/1` is useful when you might use a nested `case/2` statement or situations that cannot cleanly be piped together. The `with/1` expression is composed of the keywords, the generators, and finally an expression.
+- Each expression on the right side of the `|` character evaluates to a `Bool`
+- For the first one which evaluates to `True`, its expression is evaluated to
```
Wouldn't be nice to me more explicit: "The first pattern that matches -- and only it -- has its expression evaluated to"?

lbseale replied:
Added to latest push

why-not-try-calmer commented on en/lessons/basics/pattern-matching.md:
````diff
@@ -1,107 +1,283 @@
+```haskell
+{-# LANGUAGE OverloadedStrings #-}
+
+import Data.Text ( Text )
````
Same remark as for the previous document above: Why make Text the default instead of String?

why-not-try-calmer commented on en/lessons/basics/pattern-matching.md:
```diff
@@ -1,107 +1,283 @@
+[]       -- An empty list
+(x:xs)   -- A list with one or more values, constructed with `:`
+         -- `x` is the first element of the list, and `xs` is the rest of the list
+Nothing  -- A constructor for the `Maybe` type
```
`Maybe` is not a type. `Maybe a` is.

Kleidukos replied:
`Maybe` is very much a type, of kind `* -> *`. `Maybe a` is a type of kind `*`. Being a higher-kinded type doesn't make it less of a type.

why-not-try-calmer replied:
What I mean to say is that since `Nothing :: Maybe a`, the correct thing to say is that `Nothing` is a constructor for the `Maybe a` datatype. More generally, and to put the point in different terms, C is a constructor to a type T only if C constructs of type whose kind = the kind of T. Since `Nothing` and `Maybe` does not agree in kind, the former cannot be a constructor to the latter.

lbseale replied:
Decided to leave this as-is

why-not-try-calmer commented on en/lessons/basics/pattern-matching.md:
```diff
@@ -1,107 +1,283 @@
+  called, its expression is evaluated
+- The underscore (`_`) is a catch-all that matches any value
+
+### Patterns In General
```
I fee llike there should be something about pattern matching on records constructor here since they are so attractive to new users, perhaps in conjunction -- unless it is deemed too advanced -- with XRecordwilcards, XRecordpuns and the 'as-' operator.

Kleidukos replied:
Interesting addition, I agree.

lbseale replied:
Nice one, I'll add this

lbseale replied:
Just added this to latest push

lbseale committed:
Changes for latest comments on PR

- Add some explicitness to control structures
- Add pattern matching with records
- Add at-patterns
- Add an example of tuples

commit aed517f9d22b91711812500c9842894861631e68
Author: Luke <19484084+lbseale@users.noreply.github.com>
Date:   Wed Oct 6 07:16:16 2021 -0700

    Remove temporary source file for examples

commit d7e68e28117c09f16b68a0318974015b60c21ba8
Author: Luke <19484084+lbseale@users.noreply.github.com>
Date:   Wed Oct 6 07:14:13 2021 -0700

    Import qualified Text as T

    - I think this will be less confusing, since both the module and the
    type are already named Text

commit 0c789722a131bcbbe60d7323fd6c0277d4b168ca
Author: Luke <19484084+lbseale@users.noreply.github.com>
Date:   Wed Oct 6 06:48:15 2021 -0700

    Add simple tuple example

commit a9106a1cc1c65cfc06767eb5139072294072277a
Author: Luke <19484084+lbseale@users.noreply.github.com>
Date:   Wed Oct 6 06:47:54 2021 -0700

    Add example for at-pattern naming

commit 9931816c1dcab7a2b432882cac76ad3af4c3d452
Author: Luke <19484084+lbseale@users.noreply.github.com>
Date:   Tue Oct 5 07:05:44 2021 -0700

    Add some explicitness to control structures

commit c2bcdf8bb685945caf7b78de9b6b1ffefb1160ea
Author: Luke <19484084+lbseale@users.noreply.github.com>
Date:   Tue Oct 5 07:05:28 2021 -0700

    Add record pattern example

commit 112a142cbec44c2cf91d4ca0f0ad3ca17f2c969a
Author: Luke <19484084+lbseale@users.noreply.github.com>
Date:   Tue Oct 5 07:04:56 2021 -0700

    Add haskell source file to test examples

Kleidukos commented on en/lessons/basics/pattern-matching.md:
```diff
@@ -1,107 +1,426 @@
+   defined as a local name
+
+Feel free the explore these further once you feel confident writing record
+patterns by hand
```
A period is missing. :)

lbseale replied:
None of the one-line statements have periods. I think it looks a bit nicer this way, should I go and add periods to them?

Kleidukos replied:
I think that from a typographical point of view it's better, yes. :)

lbseale replied:
Ok just pushed with added punctuation 👍

Kleidukos reviewed (changes requested):
Nitpick. :)

lbseale committed:
Add punctuation

- Add periods everywhere
- Add colons (:) to lines refering to subsequent code blocks

Kleidukos reviewed (approved):
Well it would seem we've arrived to a satisfying point! Thank you very much @lbseale for your patience and dedication. :)
